### DISCUSSION TITLE
I should be able to use NotNullWhenAttribute on a method that returns an enum

### DISCUSSION THREAD

**Discussion by PhilipEve** [dotnet/csharplang#7012] [Language Ideas]
At the moment, you can only use NotNullWhenAttribute on methods returning bool.

I should be able to have a method that returns an enum, and specify that an out parameter is not null when the method returns certain values of the enum.

**Comment by Joe4evr** [comment]
I have run into this at some point as well. No idea if it's feasable to retrofit this functionality onto `{Member}NotNullWhenAttribute` as they exist now, but it would help some cases.

**Comment by PhilipEve** [comment]
To add slightly more justification on why enums might be a particular case where this would be desirable: if I'm returning an enum from a method and that method also has out parameters, then this is a somewhat complex method we're talking about (or it's orchestrating complex operations in other methods it calls into).

The enum return value is probably partitioning the space of possible outcomes into qualitative categories. It makes sense that you might want to emit a non-null value for the out parameter when the outcome is in some of those categories but not others. Boolean is in some sense the simplest possible case of this, the fact that Boolean isn't an enum in .NET notwithstanding.
